src/test/java/net/hydromatic/lookml/test/LaxTest.java

```diff
@@ -0,0 +1,126 @@
+                .code("code", "VALUES 1")
+                .list("list", h2 ->
+                    h2.identifier("asc")
+                        .number(-2.5)
+                        .string("abc")
+                        .list(h3 -> h3.string("singleton")))
+                .list("emptyList", h2 -> {}))
+        // This close() is spurious, but LookmlWriter currently needs it
+        // to trigger a flush of the internal state.
+        .close();
+  }
+
+  /** Tests the LookML writer
+   * {@link LaxHandlers#writer(StringBuilder, int, boolean)}
+   * by running a sequence of parse events through the writer and checking
+   * the generated LookML string. */
+  @Test void testWriter() {
+    final StringBuilder b = new StringBuilder();
+    generateSampleEvents(LaxHandlers.writer(b, 2, true));
+    final String lookml = "model: m {\n"
+        + "  n: 1\n"
+        + "  s: \"hello\"\n"
+        + "  b: true\n"
+        + "  code: VALUES 1;;\n"
+        + "  list: [\n"
+        + "    asc,\n"
+        + "    -2.5,\n"
+        + "    \"abc\",\n"
+        + "    [\n"
+        + "      \"singleton\"\n"
+        + "    ]\n"
+        + "  ]\n"
+        + "  emptyList: []\n"
+        + "}";
+    assertThat(b, hasToString(lookml));
+
+    // Same as previous, in non-pretty mode
+    b.setLength(0);
+    generateSampleEvents(LaxHandlers.writer(b, 2, false));
+    final String lookml2 = "model:m {"
+        + "n:1, s:\"hello\", b:true, code:VALUES 1;;, "
+        + "list:[asc, -2.5, \"abc\", [\"singleton\"]], "
+        + "emptyList:[]"
+        + "}";
+    assertThat(b, hasToString(lookml2));
+  }
+
+  /** Tests the logging handler
+   * {@link LaxHandlers#logger}
+   * by running a sequence of parser events through it and checking the
+   * resulting list of strings. */
+  @Test void testLogger() {
+    final List<String> list = new ArrayList<>();
+    generateSampleEvents(LaxHandlers.logger(list::add));
+    assertThat(list,
+        hasToString("[objOpen(model, m), "
+            + "number(n, 1), "
+            + "string(s, hello), "
+            + "identifier(b, true), "
+            + "code(code, VALUES 1), "
+            + "listOpen(list), "
+            + "identifier(asc), "
+            + "number(-2.5), "
+            + "string(abc), "
+            + "listOpen(), "
+            + "string(singleton), "
+            + "listClose(), "
+            + "listClose(), "
+            + "listOpen(emptyList), "
+            + "listClose(), "
+            + "objClose()]"));
+  }
+
+  /** Tests that the same events come out of a filter handler
+   * ({@link LaxHandlers#filter(ObjectHandler)}) as go into it. */
+  @Test void testFilterHandler() {
+    final List<String> list = new ArrayList<>();
+    generateSampleEvents(LaxHandlers.logger(list::add));
+    final List<String> list2 = new ArrayList<>();
+    generateSampleEvents(LaxHandlers.filter(LaxHandlers.logger(list2::add)));
+    assertThat(list2, hasSize(list.size()));
```

Comment:
If you're testing for `same events` for the lists, why bother testing for `size`?  Wouldn't the source of truth be the actual list comparison?

Reply:
You are correct that, if everything is working, then this check is superfluous.

But tests are also useful when not everything is working (e.g. in the middle of refactoring).